### PR TITLE
bulk, backupccl: introduce a Structured event Aggregator

### LINKED ISSUE #80388: bulk: implement a bulk aggregator that uses tracing to surface information about suboperations
Stemming from a conversation with @andreimatei and @dt.

A bulk aggregator is an object that is responsible for aggregating and rendering information about suboperations spawned by a bulk processor. To begin with, we will focus on aggregating two data points:

1) The duration of each operation.

2) Interesting `StructuredEvent`s that can be used to convey more detailed metrics such as sst bytes ingested, wait time per store, number of sst batcher flushes etc.

For 1) we will start by addressing #80391. This will allow us to simply ask our root tracing span for a snapshot of its `childSpan -> duration` mapping at render time. Note, the bulk aggregator will be set as a `LazyTag` on the processor's root context, which allows us to write bespoke logic in our `Render` function.

For 2) we need to build out the structured event listener in #80395. This is an important piece because we do not want the bulk aggregator to miss events as they are rotated out of the tracing span's ring buffer. Bulk jobs are known to run for hours and send thousands of RPCs per node, so it is certain that structured events will fall out of the buffer during the execution of the job. We would register the bulk aggregator as an event listener on the processor's root span, and this will allow us to intercept interesting events, and roll them up into a running aggregate.

The goal is to build a generic aggregator that can keep track of **any** suboperation that wraps its execution in a child span, without the overhead of notifying the aggregator at the various call sites. This is very attractive since threading `notify` calls in different parts of bulk code is a lot more of a lift than maintaining good hygiene about wrapping important operations in child spans. 

There is more exploratory work to be done on how we will present all the information in the bulk aggregator on the tracez page but that is outside the scope of this issue. https://github.com/cockroachdb/cockroach/pull/80198 broke some ground and we might be able to use some of the changes there.

Epic: [CRDB-10262](https://cockroachlabs.atlassian.net/browse/CRDB-10262)

Jira issue: CRDB-15736

Epic CRDB-10262

### PR DESCRIPTION
**Pull request by adityamaru**
This change introduces an Aggregator object that is capable
of listening for Structured events emitted in a recording,
aggregating them and rendering them as LazyTags.
We also introduce an AggregatorEvent interface that can be
implemented by a Structured event thereby making it eligible
for aggregation in the Aggregator.

The first user of the Aggregator will be every backup data
processor that is setup on the nodes in the cluster during
a backup. The Aggregator lives as long as the processor, and
listens for Aggregator events emitted by any span that is a child
of the processors' span. This includes both local children as
well as remote children whose recordings have been imported into
a local span. The Aggregator stores running aggregates of each
AggregatorEvent it is notified about, bucketed by the events'
tag name. This aggregate will be rendered on the tracing span as a LazyTag.

This change teaches every ExportRequest to emit an AggregatorEvent.
Going forward we expect many more operations in bulk jobs to
define and emit such events providing visibility into otherwise
opaque operations.

We cleanup some of the StructuredEvents that were previously
added but have not proved useful, and also change some of the
tracing span operation names to be more intuitive.

To view these aggregated events once can navigate to the
`/tracez` endpoint of the debug console to take a snapshot and search for either
`BACKUP` or the job ID to filter for tracing spans on that node.
The span associated with the backup processor will be decorated with
tags that correspond to the fields in the introduced `ExportStats`
proto message. Note these stats are only aggregated on a per node
basis.

Fixes: #80388

Release note: None

Release justification: low risk change for improved observability into backups.

### CODE REVIEW

**Review by andreimatei — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84043)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @rhu713, and @stevendanna)

___
*[`pkg/util/bulk/aggregator.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N6jGxGx2rAfNt6xDJqQ:-N6jGxGx2rAfNt6xDJqR:b-lc6rjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/34e7ce72f09c2e8fda312bff1dd40564348404c0/pkg/util/bulk/aggregator.go#L64)):*
> ```Go
> 		b.mu.sp.SetLazyTag(eventTag, b.mu.aggregatedEvents[eventTag])
> 	} else {
> 		b.mu.aggregatedEvents[eventTag].Combine(bulkEvent)
> ```

does this change an event that's part of the structured recording, or does it change a copy that's only used for the tag?
___
*[`pkg/util/bulk/aggregator.go` line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N6jGLDU5H39bJD9RDqE:-N6jGLDU5H39bJD9RDqF:b-5zsumi) ([raw file](https://github.com/cockroachdb/cockroach/blob/34e7ce72f09c2e8fda312bff1dd40564348404c0/pkg/util/bulk/aggregator.go#L83)):*
> ```Go
> // The Aggregator instance is responsible for finishing the returned span, and
> // so the user must call Close().
> func MakeAggregatorWithSpan(
> ```

I don't need to care about this code, so feel free to ignore, but I for one would not intertwine the aggregator and the span. I'd have callers create a span explicitly, then attach an aggregator to the span. The aggregator would not be in charge of closing the span.
___
*[`pkg/util/tracing/crdbspan.go` line 512 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N6iWafm4KjvvyHuLo0J:-N6iWafm4KjvvyHuLo0K:b-hfle69) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a165c03643ed43a0d45b16c0a795244543a2fe/pkg/util/tracing/crdbspan.go#L512)):*
> ```Go
> 	for _, span := range childRecording {
> 		for _, record := range span.StructuredRecords {
> 			s.notifyEventListeners(record.Payload)
> ```

remind me why this had to change pls

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84043)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @rhu713, and @stevendanna)

___
*[`pkg/util/bulk/aggregator.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N6jGxGx2rAfNt6xDJqQ:-N9HGzRP7Nlotk3uMkq0:b-b39vcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/34e7ce72f09c2e8fda312bff1dd40564348404c0/pkg/util/bulk/aggregator.go#L64)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

does this change an event that's part of the structured recording, or does it change a copy that's only used for the tag?
</blockquote></details>

good catch, added a Clone method to the interface that returns a copy the first time we assign it a slot in the map.
___
*[`pkg/util/bulk/aggregator.go` line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N6jGLDU5H39bJD9RDqE:-N9HCEsFDkFEldKXE4Qo:bwrocw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/34e7ce72f09c2e8fda312bff1dd40564348404c0/pkg/util/bulk/aggregator.go#L83)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't need to care about this code, so feel free to ignore, but I for one would not intertwine the aggregator and the span. I'd have callers create a span explicitly, then attach an aggregator to the span. The aggregator would not be in charge of closing the span.
</blockquote></details>

I did try this before the current implementation, but it wasn't very pretty. Since event listeners have to be registered at span creation time, it involved plumbing the event listener through a few methods in processorbase starting from `bp.StartInternal(ctx, backupProcessorName)`. 
___
*[`pkg/util/tracing/crdbspan.go` line 512 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N6iWafm4KjvvyHuLo0J:-N9HBksb2Cs-0ZGZOxiZ:b-xjs1og) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a165c03643ed43a0d45b16c0a795244543a2fe/pkg/util/tracing/crdbspan.go#L512)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

remind me why this had to change pls
</blockquote></details>

https://github.com/cockroachdb/cockroach/pull/84043/files#diff-d5042f9b360d17b5e8009b0cedd5b8a7c490b64b4b41c6958b4e5fd99074ad57R47 in the aggregator we type check that the structured event is an `AggregatorEvent`. When importing remote spans we notify the event listeners on the span being imported into of these events. Before this diff we would pass along `record.Payload` which is of type `*types.Any`. Casting `record.Payload` to the `AggregatorEvent` interface type will always fail, since it is the underlying `protoutil.Message` that implements `AggregatorEvent`. So we first unmarshal the types.Any into a types.DynamicAny which would give us access to the concrete proto type.

File: pkg/util/tracing/crdbspan.go
Comment (benbardin) [inline pkg/util/tracing/crdbspan.go:938]: Huh, didn't you fix this already?

**Review by benbardin — APPROVED** [review]

**Comment by benbardin** [comment]
Approved for backupccl

File: pkg/util/tracing/crdbspan.go
Comment (adityamaru) [reply pkg/util/tracing/crdbspan.go:938]: I don't think so, I rebased on master a couple days ago. When we chatted it was me working on this diff its just been sitting in cold storage since. TFTR! I'll wait on Steven to give it a pass since I bugged him to before you stamped it, and then merge away.

File: pkg/util/tracing/crdbspan.go
Comment (stevendanna) [inline pkg/util/tracing/crdbspan.go:516]: Perhaps it doesn't matter too much, but I kinda wonder if we can hold off on Unmarshaling this until we know there are event listeners in place for this span.

**Review by stevendanna — APPROVED** [review]
Overall LGTM, I left some comments, but I leave them to your discretion, I don't think they are blocking.

It might be nice to include in the already good commit message a line or two about how I might see these aggregated events -- just so anyone who sees this new feature knows how to start experimenting with it.

File: pkg/ccl/backupccl/backup_processor.go
Comment (adityamaru) [reply pkg/ccl/backupccl/backup_processor.go:519]: good point, let me flesh out the comment on the Duration field to capture this.

File: pkg/util/tracing/crdbspan.go
Comment (adityamaru) [reply pkg/util/tracing/crdbspan.go:516]: I think we'd either have to change the `notifyEventListeners` signature to take a `*types.Any` or inside `notifyEventListeners` we will have to `MarshalAny` the passed in`item Structured` before we UnmarshalAny it into a DynamicAny. I think I'm going to leave it as is for now since this code is only called when importing remote spans into our span, not when every structured event is emitted.

File: pkg/ccl/backupccl/backuppb/backup.go
Comment (adityamaru) [reply pkg/ccl/backupccl/backuppb/backup.go:170]: Changed Clone() to be a deep copy instead so that we don't have to worry about accidentally shallow copying some fields but not others.

File: pkg/util/bulk/aggregator.go
Comment (rhu713) [inline pkg/util/bulk/aggregator.go:63]: Wonder if we should change the interface for `AggregatorEvent` to be similar to what I've seen elsewhere e.g. in Java's reduce pattern: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#reduce-T-java.util.function.BinaryOperator-

Instead of `Clone()` we use `Identity()` with `Combine()` here, and maybe having an identity buys us more flexibility in the future.

**Review by rhu713 — COMMENTED** [review]
code LGTM. I just have a slight design input that may improve things down the line.

File: pkg/util/bulk/aggregator.go
Comment (adityamaru) [reply pkg/util/bulk/aggregator.go]: nice idea, check it out!

**Review by rhu713 — APPROVED** [review]

**Review by andreimatei — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-:-N9gKeXFB-2ls8JnZTcT:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @andreimatei, @benbardin, @dt, @rhu713, and @stevendanna)

___
*[`pkg/util/bulk/aggregator.go` line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gKnG8D0LXMPy62Byp:-N9gKnG8D0LXMPy62Byq:b-d8h3d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L23)):*
> ```Go
> // Aggregator. An AggregatorEvent also implements the tracing.LazyTag interface
> // to render its information on the associated tracing span.
> type AggregatorEvent interface {
> ```

given that there is only one type of event implementing this, is the interface worth it?
___
*[`pkg/util/bulk/aggregator.go` line 37 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gILbP3XVUuMEbxUc8:-N9gILbP3XVUuMEbxUc9:b-s54f1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L37)):*
> ```Go
> // An Aggregator can be used to aggregate and render AggregatorEvents that are
> // emitted as part of its tracing spans' recording.
> type Aggregator struct {
> ```

consider naming this `TracingEventAggregator`, or something more specific than simply `Aggregator`.
___
*[`pkg/util/bulk/aggregator.go` line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gK337AYb4y30hjHA0:-N9gK337AYb4y30hjHA1:b1lvgnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L61)):*
> ```Go
> 
> 	// If this is the first AggregatorEvent with this tag, set it as a LazyTag on
> 	// the associated tracing span. This way the AggregatorEvent will be
> ```

nit: scratch the second sentence. Every user of a tag is not expected to explain what the point of tags are.
___
*[`pkg/util/bulk/aggregator.go` line 96 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gKW3c26vCFYAZJiQb:-N9gKW3c26vCFYAZJiQc:b-codqvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L96)):*
> ```Go
> 	sp := tracing.SpanFromContext(ctx)
> 
> 	aggCtx, aggSpan := sp.Tracer().StartSpanCtx(ctx, aggregatorName,
> ```

use `EnsureChildSpan`, and then you don't need to deal with the parent.
___
*[`pkg/util/bulk/aggregator.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gHtNSDu6tvwHSv0ii:-N9gHtNSDu6tvwHSv0ij:b-gjrdsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L97)):*
> ```Go
> 
> 	aggCtx, aggSpan := sp.Tracer().StartSpanCtx(ctx, aggregatorName,
> 		tracing.WithEventListeners([]tracing.EventListener{agg}), tracing.WithParent(sp))
> ```

let's make `WithEventListener` a variadic function (take `...EventListener`)

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84043)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @benbardin, @dt, @rhu713, and @stevendanna)

___
*[`pkg/ccl/backupccl/backuppb/backup.go` line 176 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N6PkePivxXAbmxt-h13-r2-176:-N9gZyK38fdfZqRCfc-7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b36270b3ca79cb33ff7f30f3654aacb7227e196/pkg/ccl/backupccl/backuppb/backup.go#L176)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We could log the type of other here.
</blockquote></details>

Done.
___
*[`pkg/util/bulk/aggregator.go` line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gKnG8D0LXMPy62Byp:-N9gTWUM3xQADNxLddc9:b-bc1z3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L23)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

given that there is only one type of event implementing this, is the interface worth it?
</blockquote></details>

I'm bullish that there'll be more very soon. Especially in import/restore to surface AddSSTable, BulkAdder stats etc.
___
*[`pkg/util/bulk/aggregator.go` line 37 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gILbP3XVUuMEbxUc8:-N9gWUu4DZhS1KwIzEGN:b-gg60ty) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L37)):*
> TracingEventAggregator
done.
___
*[`pkg/util/bulk/aggregator.go` line 96 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gKW3c26vCFYAZJiQb:-N9gZwAt4N_85XXNOj4q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L96)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

use `EnsureChildSpan`, and then you don't need to deal with the parent.
</blockquote></details>

Done.
___
*[`pkg/util/bulk/aggregator.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N9gHtNSDu6tvwHSv0ii:-N9gZvu65Ii6niliiuM8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc36b3cd6e63d50237bf8026be6de1586fbc96ae/pkg/util/bulk/aggregator.go#L97)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

let's make `WithEventListener` a variadic function (take `...EventListener`)
</blockquote></details>

Done.
___
*[`pkg/util/bulk/aggregator_test.go` line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84043#-N6PkePivxXAbmxt-h1A-r2-60:-N9gZxFICOAJrdXJ8jtL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b36270b3ca79cb33ff7f30f3654aacb7227e196/pkg/util/bulk/aggregator_test.go#L60)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

```suggestion
	// have not imported the remote child's Recording.
```

or, I suppose, 

```suggestion
	// have not imported the remote children's Recordings.
```
</blockquote></details>

Done.

**Comment by adityamaru** [comment]
TFTRs!


bors r+

**Pull request merged**
